openhcl_boot: Serial logging for SEV-SNP

romank-msft opened the pull request:
SEV-SNP has been the only platform where serial logging
isn't supported in the boot shim.

Implement the GHCB calls and use that for serial logging
via the IOIO exits.

Validated with the latest main, both the ship and debug builds.

![image](https://github.com/user-attachments/assets/4d606eb6-f85f-41c7-8744-3989eb848d43)

jstarks commented on vm/loader/src/paravisor.rs:
```diff
@@ -322,6 +324,39 @@ where
+    // shim. A static allocation inside the binary and manipulation the C-bit
+    // from inside the shim is not possible, as the shim is mapped with 2MiB
+    // pages, not using the RWX bits for different sections of the binary.
+    // Flipping the C bit for the 2MiB will sramble the contents thus scrambling
```
Can we instead dynamically allocate inside the shim a non-identity mapped PDE+PTE+whatever, with VA < 0 to avoid any possible conflicts? This could probably be used to clean up some of our LocalMap weirdness, too.

jstarks replied:
But I don't really get why the loader needs to care about this. I think ideally we'd just allocate the necessary pages using some allocator thing. Unless the idea is that this is so early that it's hard to do that. @chris-oo was trying to improve the boot loader allocator story.

jstarks replied:
Another easy option would be to reserve the memory via some `static`s in the boot loader. At least for the "additional pages" that seems simpler, at the cost of measuring the pages unnecessarily (but as an optimization we could later look at marking all of BSS as unmeasured and accepting it+zeroing it explicitly--I think we already kind of unnecessarily zero BSS in the boot loader due to a misunderstanding/bug elsewhere.)

romank-msft replied:
> a non-identity mapped PDE+PTE+whatever, with VA < 0

Doing VA > 0 atm, will move to the upper half, makes a lot of sense to me! The allocation for the additional page tables in the IGVM was in the spur of the moment 😄 Will clean that up, thanks a lot for the feedback!!

romank-msft replied:
fixed

romank-msft commented on vm/x86/x86defs/src/snp.rs:
```diff
@@ -451,9 +462,65 @@ open_enum::open_enum! {
     }
 }
 
+#[repr(C, packed)]
```
Thanks, will remove!

romank-msft commented:
Don't approve just yet folks, will look more at the page acceptance code to make sure it matches the manuals.

jstarks commented on openhcl/virt_mshv_vtl/src/processor/snp/mod.rs:
```diff
@@ -914,7 +914,7 @@ impl UhProcessor<'_, SnpBacked> {
         tracing::trace!(?ghcb_msr, "vmgexit intercept");
 
-        match x86defs::snp::GhcbInfo(ghcb_msr.info()) {
+        match x86defs::snp::GhcbInfo(ghcb_msr.info().0) {
```
Seems like you're stripping and adding back the same type. I.e., can't we just match on `ghcb_msr.info()` now?

romank-msft replied:
Will fix this, thanks!

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -32,52 +57,392 @@ pub enum AcceptGpaError {
+    }));
+
+/// The GHCB page itself.
+static GHCB: SingleThreaded<UnsafeCell<[u8; size_of::<GhcbPage>()]>> =
```
There's no guarantee that this is 4KB aligned.

romank-msft replied:
I added `align(4096)` to `GhcbPage` page itself, and I was expecting that to guarantee the alignment. Could you explain why that won't help?

romank-msft replied:
OOOOh, I see that

jstarks replied:
We've had similar bugs in the past, where usually we get lucky and the thing ends up aligned but occasionally something gets perturbed in the link order or whatever and *boom*.

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -4,12 +4,37 @@
+use x86defs::snp::SevIoAccessInfo;
+use zerocopy::IntoBytes;
+
+static GHCB_PREVIOUS: AtomicU64 = AtomicU64::new(0);
```
Just use `SingleThreaded` to avoid needing to treat this as atomic.

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -32,52 +57,392 @@ pub enum AcceptGpaError {
+        unsafe { write_msr(X86X_AMD_MSR_GHCB, GHCB_PREVIOUS.load(Ordering::Acquire)) };
+    }
+
+    fn ghcb_mut() -> &'static mut GhcbPage {
```
This is not safe. You can call this twice and get two mut references to the same page, which is undefined behavior.

Probably you should use `off_stack!` to define these things so that you can get something kind of like a `RefCell`--it will catch at runtime if you try to get two references to it.

(I see we have the same bug in hypercall.rs. We should fix that in a followup.)

jstarks replied:
Oh, I see, that doesn't quite work. Hmm.

jstarks replied:
I think just mark this function as `unsafe` for now, so that the caller has to justify why they think there are no concurrent references.

romank-msft replied:
done, thanks a bunch!

romank-msft commented:
@jstarks Addressed your comments, appreciated very much! `pvalidate` makes the VM shut down when accepting the GHCB page back when unintializating. Looking what's up, followed the usual procedure, shuffled the steps around just in case. @chris-oo @mebersol @sluck-msft could you look to see what's suspicious?

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -32,52 +59,422 @@ pub enum AcceptGpaError {
+    #[must_use]
+    fn read_io_port(port: u16, access_size: IoAccessSize) -> Option<u32> {
+        // SAFETY: The GHCB page is statically allocated and initialized,
+        // and the GHCB page is not accessed concurrently: this code runs
```
Ah, so, this helps uncover another issue--the GHCB _is_ concurrently accessed, by a malicious host. So, we can't just use a `mut` reference to access it--the compiler can legally do weird things that could cause security issues (although it's very unlikely to do so).

So, once the GHCB page is mapped as shared, we need to use only atomic accesses on it. You can use the stuff in `safeatomic` to make this easier. In will probably look like:

```rust
fn ghcb() -> &[AtomicU8; 4096] {
    ...
}

let ghcb = Self::ghcb();
ghcb[offset_of!(Ghcb, ghcb_usage)][..4].atomic_write(GhcbUsage::BASE.0);
...
// Maybe write the whole save area in one go to simplify:
ghcb[offset_of!(Ghcb, save)][..size_of::<SaveThing>()].atomic_write(GhcbSaveArea { ... });
```

This will be annoying to do, sorry. You could add some helpers to make this easier if we expect other GHCB calls in the future.

romank-msft replied:
Happy to change the code for the better, no worries at all! Thanks for your help :)

romank-msft replied:
Coded up the atomic access!

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -32,52 +59,422 @@ pub enum AcceptGpaError {
+        pvalidate(page_number, GHCB_GVA.into_bits(), false, true).expect("memory accept");
+
+        // Needed here, not before.
+        flush_tlb();
```
Is it possible that this needs to be _before_ `pvalidate` so that the C bit is observed?

romank-msft replied:
Did before, `pvalidate` would fail. Moved over here as the kernel does. Been rummaging through manuals, would be great to have AMD's insight if you can tag somebody

jstarks commented on openhcl/openhcl_boot/src/arch/x86_64/snp.rs:
```diff
@@ -32,52 +59,422 @@ pub enum AcceptGpaError {
+        let resp = Self::ghcb_call(GhcbCall {
+            extra_data: 0,
+            page_number: ghcb_page_number(),
+            info: GhcbInfo::REGISTER_REQUEST,
```
It seems that such a mechanism wasn't even defined until Jan 2025...

You may need to use the Hyper-V hypercall over GHCB mechanism to unregister itself by issuing a set register hypercall.

romank-msft replied:
My only clue has been the unregistering, still not sure if random!

Before the "fixes" commit where I changed passing GPA instead of PFN (due to which had been no serial output), `pvalidate` worked when accepting the page (obv not registered as GHCB). After I registered the correct GHCB, `pvalidate` started failing. Again, not still sure of causation.

romank-msft replied:
That's a great idea, will try!

romank-msft replied:
From the code, it seems that the hypervisor allows unmapping the GHCB overlay only for the exo partitions. I'm building a private to see if the simple fix is enough

romank-msft replied:
That made NT explode... Anyway, a silly trick of
```rust
        // A silly trick to unmap the GHCB overlay page.
        let resp = Self::ghcb_call(GhcbCall {
            extra_data: 0,
            page_number: 0,
            info: GhcbInfo::REGISTER_REQUEST,
        });
```
made the hypervisor unmap the overlay. Later on, the kernel sets its own GHCB page so very likely no harm. Might leave a bad taste in the mouth, got no more rabbits in the hat :)

romank-msft commented:
> @jstarks Addressed your comments, appreciated very much! `pvalidate` makes the VM shut down when accepting the GHCB page back when unintializating. Looking what's up, followed the usual procedure, shuffled the steps around just in case. @chris-oo @mebersol @sluck-msft could you look to see what's suspicious?

found a way to rseolve

romank-msft committed:
openhcl_boot: Serial logging for SEV-SNP

SEV-SNP has been the only platform where serial logging
isn't supported in the boot shim.

Implement the GHCB calls and use that for serial logging
via the IOIO exits.
